otherlibs/unix/sleep.c

```diff
@@ -35,26 +35,31 @@ CAMLprim value unix_sleep(value duration)
   {
     struct timespec t;
     int ret;
-    caml_enter_blocking_section();
     t.tv_sec = (time_t) d;
     t.tv_nsec = (d - t.tv_sec) * 1e9;
     do {
+      caml_enter_blocking_section();
       ret = nanosleep(&t, &t);
+      /* MPR#7903: if we were interrupted by a signal, and this signal
+         is handled in OCaml, we should run its handler now,
+         not at the end of the full sleep duration.  Leaving the blocking
+         section and re-entering it does the job. */
+      caml_leave_blocking_section();
     } while (ret == -1 && errno == EINTR);
-    caml_leave_blocking_section();
     if (ret == -1) uerror("sleep", Nothing);
   }
 #elif defined(HAS_SELECT)
   {
     struct timeval t;
     int ret;
-    caml_enter_blocking_section();
     t.tv_sec = (time_t) d;
     t.tv_usec = (d - t.tv_sec) * 1e6;
     do {
+      caml_enter_blocking_section();
       ret = select(0, NULL, NULL, NULL, &t);
+      /* MPR#7903: same comment as above */
+      caml_leave_blocking_section();
```

Comment:
I think `errno` should be saved prior to calling leave_blocking_section, and the saved errno should be compared to EINTR below.
Otherwise if any OCaml signal handlers are executed they may overwrite errno.

Reply:
`caml_leave_blocking_section` carefully preserves `errno` in a local variable and restores it before returning, for the reason you mention.  So the code is correct as written.